### PR TITLE
Add Python input/output classes to provider codegen

### PR DESCRIPTION
**Pull request by justinvp**
This updates the Python codegen to emit input/output classes. It depends on https://github.com/pulumi/pulumi/pull/5033.

Current diffs (these are easier to review than the diffs that are automatically generated by GHA, because I baselined other codegen changes):
 - AWS: https://github.com/pulumi/pulumi-aws/commit/a77712011c3dc7577a2737b5b3004ee529ea5dc6
 - Kubernetes: https://github.com/pulumi/pulumi-kubernetes/commit/ea90ce49b0ef9b760b94be47207eb95fd7f2890b

Fixes https://github.com/pulumi/pulumi/issues/4790
Fixes https://github.com/pulumi/pulumi/issues/5140

### CODE REVIEW

**Comment by justinvp** [comment]
There are some remaining TODOs, but this can be reviewed in the meantime. I'll be finishing the "Add type annotations to Resource/Function args" today.

File: pkg/codegen/python/gen.go [hunk 1]
```diff
@@ -76,12 +84,70 @@ type modContext struct {
...
 	return title(components[2])
 }
 
+func (mod *modContext) tokenToModule(tok string) string {
+	canonicalModName := mod.pkg.TokenToModule(tok)
+	modName := PyName(strings.ToLower(canonicalModName))
+	if override, ok := mod.modNameOverrides[canonicalModName]; ok {
+		modName = override
+	}
+	return modName
```
[the remark this answers is not in the document]
Comment (justinvp) [reply pkg/codegen/python/gen.go:148]: Yes, definitely! There's room in some other places to share more code, so I'll take a look at cleaning that up as well.

**Comment by komalali** [comment]
Looking over this more closely but just wanted to answer your question about `UsesIOClasses`. The reason for setting that boolean was to avoid using input/output classes in the generated examples if we hadn't yet shipped the classes themselves for a given provider (in the case that a provider update happened after my change was merged but before yours was). 

So whenever this change is pushed out we should also update the schema and set that bool to `true`. That way the program-gen updates can happen alongside the SDK updates.

File: pkg/codegen/python/gen.go [hunk 2]
```diff
@@ -1102,96 +1435,83 @@ func (mod *modContext) genPropDocstring(w io.Writer, name string, prop *schema.P
...
+	case *schema.UnionType:
+		if !input {
+			if t.DefaultType != nil {
+				return mod.typeString(t.DefaultType, input, wrapInput, optional, acceptMapping)
+			}
+			typ = "Any"
+		} else {
+			elementTypeSet := stringSet{}
+			var elementTypes []schema.Type
+			for _, e := range t.ElementTypes {
```
[the remark this answers is not in the document]
Comment (justinvp) [reply pkg/codegen/python/gen.go:1461]: This ended up not working for `Union[str, str]` which should collapse to just `str`. I didn't have a chance to dig-in to understand why, so going to leave this for now.

**Comment by justinvp** [comment]
I added a commit that adds a `Result` suffix to function output types (matches .NET codegen) and no longer emits the `_translate_property` for these types as the keys in these dictionaries are not translated. Here's what the diff looks like for AWS: https://github.com/pulumi/pulumi-aws/commit/4e0cfc156d51668c26901f4465cc9924987244d8  (no changes for Kubernetes)

**Comment by justinvp** [comment]
@pgavlin, I added another commit to this PR that instantiates nested output types for calls to invoke. PTAL.

**Pull request merged**